### PR TITLE
enable parallel testing using pytest-xdist

### PR DESCRIPTION
**Pull request by CunliangGeng**
This PR adds the [pytest-xdist](https://pytest-xdist.readthedocs.io/en/latest/index.html) to enable parallel testing to speed up the testing. 

To make parallel testing work, some session-scoped pytest fixtures that generate data or files used by multiple test files were updated (see this [example](https://pytest-xdist.readthedocs.io/en/latest/how-to.html#making-session-scoped-fixtures-execute-only-once)).

Results:
- sequential testing: 380s
- parallel testing (8cores): 85s

Using `pytest --durations=0` can detect the most time consuming tests. It shows that tests on downloaders are most expensive, and ignoring them can reduce testing time from 85s to 15s!

### CODE REVIEW

**Comment by CunliangGeng** [comment]
Current dependencies on/for this PR:
* `dev`
  * **PR #178** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/178?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #181** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/181?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #182** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/182?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #183** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #184** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/184?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #185** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/185?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #186** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/186?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #187** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/187?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #188** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/188?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #189** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/189?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                      * **PR #190** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/190?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
                        * **PR #191** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/191?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                          * **PR #192** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/192?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                            * **PR #193** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/193?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                              * **PR #194** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/194?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                * **PR #196** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/196?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                  * **PR #197** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/197?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                    * **PR #198** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/198?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                      * **PR #199** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/199?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                        * **PR #200** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/200?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                          * **PR #201** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/201?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                            * **PR #202** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/202?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                                * **PR #195** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/195?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/190?utm_source=stack-comment).

**Comment by CunliangGeng** [comment]
Current dependencies on/for this PR:
* dev
  * **PR #178** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/178?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #181** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/181?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #182** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/182?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #183** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #184** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/184?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #185** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/185?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #186** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/186?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #187** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/187?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #188** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/188?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #189** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/189?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                      * **PR #190** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/190?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/190?utm_source=stack-comment).

File: tests/conftest.py [hunk 1]
```diff
@@ -1,19 +0,0 @@
...
-
-@pytest.fixture
-def collection_from_file() -> StrainCollection:
-    filename = DATA_DIR / STRAIN_MAPPINGS_FILENAME
-    sut = StrainCollection.read_json(filename)
-    return sut
-
-
-@pytest.fixture
-def strain() -> Strain:
```
Comment (gcroci2) [inline tests/conftest.py:16]: Why this fixture wasn't fine in one file only?

**Review by gcroci2 — APPROVED** [review]
Just for me to understand, reading from the [docs](https://pytest-xdist.readthedocs.io/en/latest/how-to.html#making-session-scoped-fixtures-execute-only-once), pytest-xdist can be used in cases a fixture needs to execute exactly once per test session, for example specifying scope="session" in the fixture and by reading from a FileLock to produce the fixture data only once. I don't see here any .lock file though. So I wonder how the tests can run in parallel with the current edits.

File: tests/conftest.py (hunk 1 above)
Comment (CunliangGeng) [reply tests/conftest.py:16]: It's fine actually.  It's used only in two files and I moved it to the two independent files, probably making it easier to read and check.

**Comment by CunliangGeng** [comment]
> Just for me to understand, reading from the [docs](https://pytest-xdist.readthedocs.io/en/latest/how-to.html#making-session-scoped-fixtures-execute-only-once), pytest-xdist can be used in cases a fixture needs to execute exactly once per test session, for example specifying scope="session" in the fixture and by reading from a FileLock to produce the fixture data only once. I don't see here any .lock file though. So I wonder how the tests can run in parallel with the current edits.

We solved the issue in a different way. See the example below:

We used `scope="session"`, it cannot make sure that the data is generated once, but it ensure that the data is generated once in one process (or one cpu core).
We also used `tmp_path_factory` (it's [a session-level fixture](https://docs.pytest.org/en/7.1.x/reference/reference.html#tmp-path-factory) provided by pytest, unlike the function-level `tmp_path`). It will make sure the data generated in different sessions will be stored in different temporary paths. 

So for the example below, 
1. If the data is generated once for all processes, very good then
2. if the data is generated more than once for all processes, then our setting can make sure the different processes have their own copy of data in a different temporary path. Say, process 1's data is in path1, while process 2's data is in path2, making sure process 1 and process 2 are independent tests.

```
class TestMibigBGCLoader:
    @pytest.fixture(scope="session") 
    def data_dir(self, tmp_path_factory): 
        # get the temp directory shared by all workers
        download_root = tmp_path_factory.mktemp("download")
        extract_path = tmp_path_factory.mktemp("metadata")
        download_and_extract_mibig_metadata(download_root, extract_path)
        yield str(extract_path)
```

**Comment by CunliangGeng** [comment]
### Merge activity

* **Dec 19, 8:24 AM**: @@CunliangGeng started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/190).
* **Dec 19, 8:25 AM**: @@CunliangGeng merged this pull request with [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/190).

**Comment by hechth** [comment]
@florian-huber we could consider using this also for matchms to speed up test execution

**Pull request merged**